geode-for-redis/src/main/java/org/apache/geode/redis/internal/eventing/BlockingCommandListener.java

```diff
@@ -0,0 +1,121 @@
+import org.apache.geode.redis.internal.commands.RedisCommandType;
+import org.apache.geode.redis.internal.commands.executor.RedisResponse;
+import org.apache.geode.redis.internal.data.RedisKey;
+import org.apache.geode.redis.internal.netty.Coder;
+import org.apache.geode.redis.internal.netty.ExecutionHandlerContext;
+
+public class BlockingCommandListener implements EventListener {
+
+  private final ExecutionHandlerContext context;
+  private final Command command;
+  private final List<RedisKey> keys;
+  private final double timeoutSeconds;
+  private final long timeSubmitted;
+  private final AtomicBoolean active = new AtomicBoolean(true);
+
+  /**
+   * Constructor to create an instance of a BlockingCommandListener in response to a blocking
+   * command. When receiving a relevant event, blocking commands simply resubmit the command
+   * into the Netty pipeline.
+   *
+   * @param context the associated ExecutionHandlerContext
+   * @param command the blocking command associated with this listener
+   * @param keys the list of keys the command is interested in
+   * @param timeoutSeconds the timeout for the command to block in seconds
+   */
+  public BlockingCommandListener(ExecutionHandlerContext context, Command command,
+      List<RedisKey> keys, double timeoutSeconds) {
+    this.context = context;
+    this.command = command;
+    this.timeoutSeconds = timeoutSeconds;
+    this.keys = Collections.unmodifiableList(keys);
+    timeSubmitted = System.nanoTime();
+  }
+
+  @Override
+  public List<RedisKey> keys() {
+    return keys;
+  }
+
+  @Override
+  public EventResponse process(RedisCommandType commandType, RedisKey key) {
+    if (!keys.contains(key)) {
+      return EventResponse.CONTINUE;
+    }
+
+    if (active.compareAndSet(true, false)) {
+      resubmitCommand();
+    }
+    return EventResponse.REMOVE_AND_STOP;
+  }
+
+  @Override
+  public void resubmitCommand() {
+    // Recalculate the timeout since we've already been waiting
+    double adjustedTimeoutSeconds = 0;
+    if (timeoutSeconds > 0.0D) {
+      long timeoutNanos = (long) (timeoutSeconds * 1e9);
+      long adjustedTimeoutNanos = timeoutNanos - (System.nanoTime() - timeSubmitted);
+      adjustedTimeoutNanos = Math.max(1, adjustedTimeoutNanos);
+      adjustedTimeoutSeconds = ((double) adjustedTimeoutNanos) / 1e9;
+    }
+
+    // The commands we are currently supporting all have the timeout at the end of the argument
+    // list. Some newer Redis 7 commands (BLMPOP and BZMPOP) have the timeout as the first argument
+    // after the command. We'll need to adjust this once those commands are supported.
+    List<byte[]> commandArguments = command.getCommandArguments();
+    commandArguments.set(commandArguments.size() - 1, Coder.doubleToBytes(adjustedTimeoutSeconds));
+
+    context.resubmitCommand(command);
+  }
+
+  @Override
+  public void scheduleTimeout(ScheduledExecutorService executor, EventDistributor distributor) {
+    if (timeoutSeconds == 0) {
```

Comment:
is it worth also returning early if active is false?

Reply:
No harm in doing so.